src/ir_def/check_ir.ml

```diff
@@ -488,6 +488,15 @@ let rec check_exp env (exp:Ir.exp) : unit =
       typ exp1 <: T.nat;
       typ exp2 <: T.nat;
       T.bool <: t
+    | SameReference, [exp1; exp2] ->
+      let t1, t2 = typ exp1, typ exp2 in
+      t1 <: t2;
+      t2 <: t1; (* i.e. t1 = t2 *)
+      T.bool <: t
+    | SameVariantTag tv, [exp1; exp2] ->
+      typ exp1 <: tv;
```

Comment:
I guess we should `assert` that `tv` is a variant type here.

Reply:
Yeah, I think so. It's a bit of a shame this isn't actually something you can encode in the type, but there's no common upper bound for variants as there is for records, right, just a lower one (and vice versa).

Reply:
Can't we say `{#} <: type exp1` etc.?